mobile/library/common/api/external.cc

```diff
@@ -15,17 +17,18 @@ static absl::flat_hash_map<std::string, void*> registry_{};
 // TODO(goaway): To expose this for general usage, it will need to be made thread-safe. For now it
 // relies on the assumption that usage will occur only as part of Engine configuration, and thus be
 // limited to a single thread.
-void registerApi(std::string name, void* api) {
+void registerApi(absl::string_view name, void* api) {
   RELEASE_ASSERT(api != nullptr, "cannot register null API");
-  registry_[name] = api;
+  registry_[std::string(name)] = api;
```

Comment:
This creates a copy. Should we pass by const reference instead?

Reply:
we can, but string_view does the same (takes the reference from the caller instead).  Net result would be one copy of the string in both situations.

ditto for below

Reply:
the other option was for me to keep this as std::string, and do registry_[std::move(name)], which I considered, but then in the retrieveApi function below, the `name` param is used twice, so a move wouldn't work there, and I preferred consistency in the APIs.

Reply:
I believe calling `std::string(name)` will create a copy. That's what I meant in an earlier comment. Sorry, if it wasn't clear.

If you do `registry_[std::move(name)]`, you need to make sure the caller must also do a move when calling `registerApi(std::move(name), ...)` unless it's an rvalue. So, IMHO it's probably not worth doing a move. Furthermore, we probably want to be explicit with using `&&` as per the style guide if we want do a move.

When passing a const reference, I believe there's no copy involved when passing it into `register_[name]`.
```
void registerApi(const std::string& name, void* api) {
  registry_[name] = api; // no copy of the string `name`.
  ...
}
```

Anyway, this is probably more like a nit :) It probably doesn't matter in a real world scenario.

Reply:
> When passing a const reference, I believe there's no copy involved when passing it into register_[name].

Doesn't there have to be a copy, because registry_ is going to store a copy of name in the flat hash map?

Reply:
> > When passing a const reference, I believe there's no copy involved when passing it into register_[name].
> 
> Doesn't there have to be a copy, because registry_ is going to store a copy of name in the flat hash map?

Yes exactly, `registery_[name] = api`, even if `name` is a `const std::string&`, will create a copy.

Reply:
What this PR does is make it one copy instead of potentially two copies

Reply:
I think I'd be inclined to go with the string&& solution and require the caller to move() in the value. If they have an std::string, it's obviously not a copy. if they don't have a string, they'll have to new one up, which will obviously be a copy. So i think this is both the most performant and the most clear?

> but then in the retrieveApi function below, the name param is used twice, so a move wouldn't work there, and I preferred consistency in the APIs.

I thought that flat_hash_map provided heterogenous lookup, so that operator[] should work with string_view instead of string. Does that work?

Reply:
ah yes, you're right about flat_hash_map keys: https://abseil.io/docs/cpp/guides/container#heterogeneous-lookup

Reply:
AH, sweet. So maybe string&& for set(), and string_view for get()?

Reply:
I like @RyanTheOptimist argument about the caller having more clarity when the type is `string&&`.  I think retrieveApi can stay `string_view` and registerApi can use `string&&`.  In a sense, it's leaking implementation details (the fact that we're using flat_hash_map where `string&&` at write vs. `string_view` at read makes sense). 
 But perhaps this is an API model we should keep for other Envoy Mobile APIs: string&& for write, string_view for reads ?

Reply:
I was drafting my reply before I saw yours Ryan, yeah I agree with that, and I think we can use that model for other EM APIs going forward for consistency: setting takes a string&&, getting takes a string_view